### ISSUE TITLE
PTZOptics (and probably all other VISCA cameras) lens speed control

### ISSUE THREAD

**Issue by mark-epstein** [bitfocus/companion-module-ptzoptics-visca#18]
<B>companion-module-ptzoptics-visca</B> currently supports "standard" lens speed, which makes it very hard to precisely set zoom and focus.
![image](https://user-images.githubusercontent.com/80222240/115651943-3e8ec600-a2fa-11eb-9aaa-1c4dab73d1cd.png)

My request is to implement the variable commands instead:
![image](https://user-images.githubusercontent.com/80222240/115652029-6b42dd80-a2fa-11eb-88c3-c96182a024de.png)

and a mechanism similar to pan/tilt to set the speed
![image](https://user-images.githubusercontent.com/80222240/115652149-aba25b80-a2fa-11eb-8708-556329125a93.png)

But as long as I'm here I'll ask that speed for pan, tilt, zoom, and focus should be options on those controls rather than just a separate setting.
![image](https://user-images.githubusercontent.com/80222240/115652295-01770380-a2fb-11eb-9cc2-33d4f02c2f43.png)

**Comment by Gartom** [comment]
@mark-epstein, I think that your request is reasonable and good. I would personally still want to have the ability to set a "master speed setting" to be used as one of the options for PTZ commands.

It may be good to know that the functionality to send generic commands to PTZOptics cameras gives some freedom to implement this in your setup already now. I use it for manual focus like this:
![bild](https://user-images.githubusercontent.com/4034632/116845355-dedfc700-abe5-11eb-9454-4c885d0f3301.png)

**Comment by mark-epstein** [comment]
> @mark-epstein, I think that your request is reasonable and good. I would personally still want to have the ability to set a "master speed setting" to be used as one of the options for PTZ commands.
> 
> It may be good to know that the functionality to send generic commands to PTZOptics cameras gives some freedom to implement this in your setup already now. I use it for manual focus like this:

Agreed, I actually do that too and have even added some acceleration if the button is held but it's annoying to enter VISCA commands
![image](https://user-images.githubusercontent.com/80222240/116876713-e2298180-abea-11eb-8eaf-42c311011261.png)

**Comment by Gartom** [comment]
Agreed, it is a bit annoying to use VISCA commands and also difficult to troubleshoot and improve a setup.

**Comment by Gartom** [comment]
I am looking into the code to see if I could find out a smart way to implement the requested support without too much effort. 

I also found that the function called "P/T Slow Mode" has been in the module from the beginning but does not seem to be implemented at all. Might be good to know.
![bild](https://user-images.githubusercontent.com/4034632/116917169-97415580-ac4e-11eb-8213-19b5f3dcbf27.png)

**Comment by Gartom** [comment]
I have started to prototype the function and I believe it will will be something like this for zoom:
![bild](https://user-images.githubusercontent.com/4034632/116940570-6cff9000-ac6e-11eb-8a47-54913de042de.png)

The idea is to set a default zoom speed and that will be used as the "Current setting" when a zoom operation is selected. The speed setting can be overridden for the particular button by using the dropdown.

**Comment by mark-epstein** [comment]
👍 That would be very nice

**Comment by Gartom** [comment]
@mark-epstein, I have implemented zoom and focus speed setting on a branch here:
https://github.com/Gartom/companion-module-ptzoptics-visca/tree/Speed-settings
Try it out if you want.

I have not tested the implementation with any cameras yet so there may be bugs and the code is definetely not ready for merge yet. Maybe I will try it on some cameras tomorrow.

**Comment by mark-epstein** [comment]
I am out of town for a few weeks but will try as soon as I can!

**Comment by nottooloud** [comment]
@mark-epstein, I tried your excellent acceleration idea, but the camera ignores the release action focus stop. Can't figure it out.

It reminded me that I always forget about stacking commands, though, so I made a nice tweak to the fixed speeds that I do have. Set the focus mode to manual as a first action on the manual focus buttons.

**Comment by Gartom** [comment]
> I tried your excellent acceleration idea, but the camera ignores the release action focus stop. Can't figure it out.

@nottooloud, did you try the code that I checked in or something else? I have not yet tried the code myself.

**Comment by nottooloud** [comment]
@Gartom, I was referring to @mark-epstein's VISCA commands. The speeds change, but it eventually free-runs. 

I'm an end user, not a coder, so I haven't tried compiling your branch.

**Comment by Gartom** [comment]
Ok, @nottooloud, did you insert a "focus stop" command to be executed on button release?
![Screenshot_20210725-212529_Adobe Acrobat.jpg](https://user-images.githubusercontent.com/4034632/126910989-6520c87e-e00b-4e4c-8f34-3f0824533242.jpg)

**Comment by nottooloud** [comment]
Yes, absolutely. It works fine with a single move on press and stop on release. When I add additional delayed faster move commands, the stop on release fails. I'm currently using 2 sets of single speed buttons, slowest and medium.

**Comment by Gartom** [comment]
@nottooloud, sorry for the late reply, but if you have a delayed action it needs to be aborted when the button is released. Use "internal: Abort actions on button". Otherwise, the delayed action may happen after you release the button.
![bild](https://user-images.githubusercontent.com/4034632/133998353-7468b84a-e714-4ff3-94ce-015bb7f9e7ca.png)

**Comment by nottooloud** [comment]
That sounds like exactly the issue. Thanks, I'll try it.

**Comment by BlakesHeaven** [comment]
Hi @Gartom, I was about to raise an issue, but looking through existing issues I found this; which is part way to what I need.
Do you know if it was ever implemented? 
I guess not as it is 21 commits behind the master branch...

It would be great to see this merged in as it is better than using command options in itself.
Would it be possible to add in the option to take a variable for the speed?

I use the command options, "81 01 04 07 2**x** FF" and "81 01 04 07 3**x** FF" to zoom in and out because the standard zoom is just too fast when controlling a cam that is going out live. I have **x** set as 0 in the StreamDeck.

I have a button that toggles between "Slow" and "Fast" that controls the PT speed and I wish to use it to also control the Zoom.

The issue I have is that using a variable breaks it, for example. ```81 01 04 07 2$(internal:custom_Zoom_Speed) FF``` 

I am happy to raise a fresh issue, but as your existing branch has most of it already it seems sensible to progress this one.

May I also make a further suggestion? The example above to ramp up the speed is great and I think a tick box to "Ramp up to full speed from 0" would be the icing on the cake. For example, if the variable/speed is 5 and the ramp-up option is enabled, it would automatically start from 0 and every 200ms increase the speed up to 5 making. Again it would be great to be able to turn this on/off via a variable.

Many thanks.
